src/value.jl

```diff
@@ -1,145 +1,194 @@
     else
-        @printf(io, "%s %s",
-                typeof(x),
-                repr(index(x.pool)[x.level]))
+        @printf(io, "%s %s", typeof(x), repr(x))
     end
 end
 
-Base.print(io::IO, x::CategoricalValue) = print(io, get(x))
-Base.repr(x::CategoricalValue) = repr(get(x))
+Base.print(io::IO, x::CatValue) = print(io, get(x))
+Base.repr(x::CatValue) = repr(get(x))
 
-@inline function Base.:(==)(x::CategoricalValue, y::CategoricalValue)
+@inline function Base.:(==)(x::CatValue, y::CatValue)
     if x.pool === y.pool
         return x.level == y.level
     else
-        return index(x.pool)[x.level] == index(y.pool)[y.level]
+        return get(x) == get(y)
     end
 end
 
-Base.:(==)(::CategoricalValue, ::Null) = null
-Base.:(==)(::Null, ::CategoricalValue) = null
+Base.:(==)(::CatValue, ::Null) = null
+Base.:(==)(::Null, ::CatValue) = null
 
 # To fix ambiguities with Base
-Base.:(==)(x::CategoricalValue, y::WeakRef) = index(x.pool)[x.level] == y
-Base.:(==)(x::WeakRef, y::CategoricalValue) = y == x
+Base.:(==)(x::CatValue, y::WeakRef) = get(x) == y
+Base.:(==)(x::WeakRef, y::CatValue) = y == x
 
-Base.:(==)(x::CategoricalValue, y::AbstractString) = index(x.pool)[x.level] == y
-Base.:(==)(x::AbstractString, y::CategoricalValue) = y == x
+Base.:(==)(x::CatValue, y::AbstractString) = get(x) == y
+Base.:(==)(x::AbstractString, y::CatValue) = y == x
 
-Base.:(==)(x::CategoricalValue, y::Any) = index(x.pool)[x.level] == y
-Base.:(==)(x::Any, y::CategoricalValue) = y == x
+Base.:(==)(x::CatValue, y::Any) = get(x) == y
+Base.:(==)(x::Any, y::CatValue) = y == x
 
-@inline function Base.isequal(x::CategoricalValue, y::CategoricalValue)
-    if x.pool === y.pool
-        return x.level == y.level
+@inline function Base.isequal(x::CatValue, y::CatValue)
+    if pool(x) === pool(y)
+        return level(x) == level(y)
     else
-        return isequal(index(x.pool)[x.level], index(y.pool)[y.level])
+        return isequal(get(x), get(y))
     end
 end
 
-Base.isequal(x::CategoricalValue, y::Any) = isequal(index(x.pool)[x.level], y)
-Base.isequal(x::Any, y::CategoricalValue) = isequal(y, x)
+Base.isequal(x::CatValue, y::Any) = isequal(get(x), y)
+Base.isequal(x::Any, y::CatValue) = isequal(y, x)
 
-Base.isequal(::CategoricalValue, ::Null) = false
-Base.isequal(::Null, ::CategoricalValue) = false
+Base.isequal(::CatValue, ::Null) = false
+Base.isequal(::Null, ::CatValue) = false
 
-Base.in(x::CategoricalValue, y::Any) = index(x.pool)[x.level] in y
-Base.in(x::CategoricalValue, y::Set) = index(x.pool)[x.level] in y
-Base.in(x::CategoricalValue, y::Range{T}) where {T<:Integer} = index(x.pool)[x.level] in y
+Base.in(x::CatValue, y::Any) = get(x) in y
+Base.in(x::CatValue, y::Set) = get(x) in y
+Base.in(x::CatValue, y::Range{T}) where {T<:Integer} = get(x) in y
 
-Base.hash(x::CategoricalValue, h::UInt) = hash(index(x.pool)[x.level], h)
-
-function Base.isless(x::CategoricalValue{S}, y::CategoricalValue{T}) where {S, T}
```

Comment:
In fact, this method is related to this PR since in this PR the comparisons could be done between different categorical value implementations.
So this method could have been kept, but to cover all cases a bunch of `CategoricalValue` vs `CategoricalString` comparisons have to be added as well.
And yet it would not cover all statically resolvable `pool(x) !== pool(y)` cases since the reference types also have to be taken into account.  

Luckily, the compiler seems to be smart enough to reduce the common `isless()` to just throwing an exception.
```
julia> z = CategoricalVector([1, 2, 3], ordered=true)
3-element CategoricalArrays.CategoricalArray{Int64,1,UInt32}:
 1
 2
 3

julia> y = CategoricalVector{String, UInt8}(["b", "a", "b"], ordered=true)
3-element CategoricalArrays.CategoricalArray{String,1,UInt8}:
 "b"
 "a"
 "b"

julia> x = CategoricalVector{String, UInt32}(["b", "a", "b"], ordered=true)
3-element CategoricalArrays.CategoricalArray{String,1,UInt32}:
 "b"
 "a"
 "b"

julia> @code_warntype Base.isless(x[1], y[1])
Variables:
  #self#::Base.#isless
  x::Any
  y::Any

Body:
  begin  # line 160:
      return (CategoricalArrays.throw)($(Expr(:new, :(Base.ArgumentError), "CategoricalValue objects with different pools cannot be tested for order")))::Union{} # line 162:
  end::Union{}

julia> @code_warntype Base.isless(x[1], z[1])
Variables:
  #self#::Base.#isless
  x::Any
  y::Any

Body:
  begin  # line 160:
      return (CategoricalArrays.throw)($(Expr(:new, :(Base.ArgumentError), "CategoricalValue objects with different pools cannot be tested for order")))::Union{} # line 162:
  end::Union{}

julia> @code_warntype Base.isless(y[1], z[1])
Variables:
  #self#::Base.#isless
  x::Any
  y::Any

Body:
  begin  # line 160:
      return (CategoricalArrays.throw)($(Expr(:new, :(Base.ArgumentError), "CategoricalValue objects with different pools cannot be tested for order")))::Union{} # line 162:
  end::Union{}

julia> @code_warntype Base.isless(y[1], y[2])
Variables:
  #self#::Base.#isless
  x::CategoricalArrays.CategoricalString{UInt8}
  y::CategoricalArrays.CategoricalString{UInt8}

Body:
  begin 
      unless (Core.Intrinsics.not_int)(((Core.getfield)(x::CategoricalArrays.CategoricalString{UInt8}, :pool)::CategoricalArrays.CategoricalPool{String,UInt8,CategoricalArrays.CategoricalString{UInt8}} === (Core.getfield)(y::CategoricalArrays.CategoricalString{UInt8}, :pool)::CategoricalArrays.CategoricalPool{String,UInt8,CategoricalArrays.CategoricalString{UInt8}})::Bool)::Bool goto 4 # line 160:
      return (CategoricalArrays.throw)($(Expr(:new, :(Base.ArgumentError), "CategoricalValue objects with different pools cannot be tested for order")))::Union{}
      4:  # line 162:
      $(Expr(:inbounds, false))
      # meta: location /home/astukalov/.julia/v0.6/CategoricalArrays/src/value.jl order 52
      SSAValue(3) = (Core.getfield)((Core.getfield)(x::CategoricalArrays.CategoricalString{UInt8}, :pool)::CategoricalArrays.CategoricalPool{String,UInt8,CategoricalArrays.CategoricalString{UInt8}}, :order)::Array{UInt8,1}
      # meta: location multidimensional.jl getindex 245
      SSAValue(5) = (Base.zext_int)(Int64, (Core.getfield)(x::CategoricalArrays.CategoricalString{UInt8}, :level)::UInt8)::Int64
      # meta: pop location
      # meta: pop location
      $(Expr(:inbounds, :pop))
      SSAValue(6) = (Base.arrayref)(SSAValue(3), SSAValue(5))::UInt8
      $(Expr(:inbounds, false))
      # meta: location /home/astukalov/.julia/v0.6/CategoricalArrays/src/value.jl order 52
      SSAValue(0) = (Core.getfield)((Core.getfield)(y::CategoricalArrays.CategoricalString{UInt8}, :pool)::CategoricalArrays.CategoricalPool{String,UInt8,CategoricalArrays.CategoricalString{UInt8}}, :order)::Array{UInt8,1}
      # meta: location multidimensional.jl getindex 245
      SSAValue(2) = (Base.zext_int)(Int64, (Core.getfield)(y::CategoricalArrays.CategoricalString{UInt8}, :level)::UInt8)::Int64
      # meta: pop location
      # meta: pop location
      $(Expr(:inbounds, :pop))
      return (Base.ult_int)(SSAValue(6), (Base.arrayref)(SSAValue(0), SSAValue(2))::UInt8)::Bool
  end::Bool
```

Reply:
OK, good to know the compiler is that smart!